### PR TITLE
Addressable led

### PR DESCRIPTION
**Pull request by 22avigada**

### CODE REVIEW

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    public void shooterError(double limelightError) {
+        double ledProportion = Math.abs(limelightError / MAX_LIMELIGHT_DEGREE);
+        int ledOn = (int) (ledProportion * MIDDLE_LED);
+        if (ledOn > MIDDLE_LED) {
+            ledOn = MIDDLE_LED;
+        }
+        int differenceInLED = MIDDLE_LED;
+        if (limelightError < 0) {
+            differenceInLED = MIDDLE_LED - ledOn;
```
Comment (ashley-wei) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java:128]: change differenceInLED variable name

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            // Set the value
+            m_ledBuffer.setHSV(i, hue, 255, 128);
+        }
+        // Increase by to make the rainbow "move"
+        m_rainbowFirstPixelHue += 3;
+        // Check bounds
+        m_rainbowFirstPixelHue %= 180;
+    }
+
+    public void gosLED() {
```
Comment (pjreiniger) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java]: What is the difference between `gosLED` and `gosColorTest`

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        generalLED(0, NUMBER_LED / 2, 255, 0, 0);
+        generalLED(NUMBER_LED / 2, NUMBER_LED, 0, 0, 255);
+    }
+
+    //1st 4 blue, next 6 red, next 10 green, next 1 yellow, the rest purple
+    public void generalLEDTest() {
```
Comment (ashley-wei) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java]: can we get rid of this

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    public void gosColorTest() {
+        for (var i = 0; i < NUMBER_LED; i += 2) {
+            m_ledBuffer.setRGB(i, 255, 0, 0);
+        }
+        for (var i = 1; i < NUMBER_LED; i += 2) {
+            m_ledBuffer.setRGB(i, 255, 255, 255);
+        }
+    }
+
+    public void singlePixel() {
```
Comment (pjreiniger) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java]: `singlePixel` makes me thing you are turning one pixel on. Maybe `movingPixel`?

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java (hunk 3 above)
Comment (ashley-wei) [reply OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java]: can we get rid of the whole generalLEDTest

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            m_pixelCounter = 0;
+        }
+    }
+
+    public void clear() {
+        generalLED(0, NUMBER_LED, 0, 0, 0);
+    }
+
+    @SuppressWarnings("PMD")
+    public void fade() {
```
Comment (pjreiniger) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java:102]: I might call this `fadeGreen` or something. Lets a caller know it only works with one color in its current state

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        for (var i = 0; i < NUMBER_LED; i += 2) {
+            m_ledBuffer.setRGB(i, 255, 0, 0);
+        }
+        for (var i = 1; i < NUMBER_LED; i += 2) {
+            m_ledBuffer.setRGB(i, 255, 255, 255);
+        }
+    }
+
+    public void singlePixel() {
+        generalLED(m_pixelCounter, m_pixelCounter + 1, 255, 0, 0);
```
Comment (pjreiniger) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java]: I think calling the generic function might be more confusing than just manually setting a single LED. The utility of the function is gone if you aren't turning off a whole bunch

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java (hunk 3 above)
Comment (pjreiniger) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java]: We can probably delete this one now

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    private static final int NUMBER_LED = 60;
+    private static final double MAX_LIMELIGHT_DEGREE = 10.0;
+    private static final int NUM_LIMELIGHT_LED = 40;
+    private static final int MIDDLE_LED = NUM_LIMELIGHT_LED / 2;
+    private final AddressableLEDBuffer m_ledBuffer;
+    private int m_rainbowFirstPixelHue;
+    private int m_pixelCounter;
+    private int m_loopCounter;
+    private double m_fadeCounter;
+    private double m_direction = 0.02;
```
Comment (pjreiniger) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java]: Maybe add something to tell that this one belongs to single pixel

**Review by pjreiniger — CHANGES_REQUESTED** [review]

File: OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+        }
+
+        if (differenceInLED < MIDDLE_LED) {
+            generalLED(differenceInLED, MIDDLE_LED, 255, 0, 0);
+        }
+        if (differenceInLED > MIDDLE_LED) {
+            generalLED(MIDDLE_LED, differenceInLED, 0, 0, 255);
+        }
+
+        System.out.println(ledProportion + "," + ledOn + "," + MIDDLE_LED + "," + differenceInLED);
```
Comment (pjreiniger) [inline OutreachBot/src/main/java/frc/robot/subsystems/LEDSubsystem.java]: Remove debug

**Review by pjreiniger — APPROVED** [review]

**Pull request merged**